File: HT_breaks.R
library(tidyverse)

HT_breaks <- function(x, k) {
  if (k > 1) {
    x0 <- vector("list", k)
    xm0 <- vector("list", (k-1))
    x0[[1]] <- x
    for (i in 1:(k-1)) {
      xm0[[i]] <- mean(x0[[i]])
      x0[[(i+1)]] <- x0[[i]][x0[[i]] > xm0[[i]]]
      if (length(x0[[(i+1)]]) <= 2) break
    }
    xbid <- c(min(x), unlist(xm0), max(x))
    xmem <- as.integer(cut(x,  breaks = xbid, label = c(1:k), include.lowest = TRUE))
    xsize <- as.integer(table(xmem))
    return(
      list(
        bin = xbid,
        size = xsize,
        member = xmem
      )
    )
  } else {
    return(NULL)
  }
}

File: HT_breaks.py
import numpy as np
import pandas as pd

def HT_breaks(x, k):
    if k > 1:
        x0 = [x]
        xm0 = []
        for i in range(k-1):
            xm0.append(np.mean(x0[i]))
            x0.append(x0[i][x0[i] > xm0[i]])
            if len(x0[i+1]) <= 2:
                break
        xbid = [np.min(x)] + xm0 + [np.max(x)]
        xmem = pd.cut(x, bins=xbid, labels=list(range(1, k+1)), include_lowest=True).astype(int)
        xsize = xmem.value_counts().to_numpy()
        return {'bin': xbid, 'size': xsize, 'member': xmem}
    else:
        return None
